pkg/controller/sidecarterminator/sidecar_terminator_controller.go

```diff
@@ -0,0 +1,360 @@
+		return reconcile.Result{}, nil
+	} else if !errors.IsNotFound(err) {
+		klog.V(3).Infof("SidecarTerminator -- Error occurred when try to get CRR(%s/%s), error: %v", existingCRR.Namespace, existingCRR.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	sidecars, _ := getSidecarNames(pod)
+	uncompletedSidecars := filterOutUncompletedSidecars(pod, sidecars)
+	var sidecarContainers []appsv1alpha1.ContainerRecreateRequestContainer
+	for _, name := range uncompletedSidecars.List() {
+		sidecarContainers = append(sidecarContainers, appsv1alpha1.ContainerRecreateRequestContainer{
+			Name: name,
+		})
+	}
+
+	crr := &appsv1alpha1.ContainerRecreateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: pod.Namespace,
+			Name:      getCRRName(pod),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(pod, pod.GroupVersionKind()),
+			},
+		},
+		Spec: appsv1alpha1.ContainerRecreateRequestSpec{
+			PodName:    pod.Name,
+			Containers: sidecarContainers,
+			Strategy: &appsv1alpha1.ContainerRecreateRequestStrategy{
+				FailurePolicy: appsv1alpha1.ContainerRecreateRequestFailurePolicyIgnore,
+			},
+		},
+	}
+
+	err = r.Create(context.TODO(), crr)
+	if err != nil {
+		klog.V(3).Infof("SidecarTerminator -- Error occurred when creating CRR(%v/%v), error %v", crr.Namespace, crr.Name, err)
+	} else {
+		klog.V(3).Infof("SidecarTerminator -- Creating CRR(%v/%v) successfully", crr.Namespace, crr.Name)
+	}
+
+	return reconcile.Result{}, err
+}
+
+func getCRRName(pod *corev1.Pod) string {
+	return fmt.Sprintf("%v-%v", pod.Name, pod.UID)
+}
+
+func allSidecarsHaveBeenCompleted(pod *corev1.Pod) bool {
+	sidecars, _ := getSidecarNames(pod)
+	for i := range pod.Status.ContainerStatuses {
+		status := &pod.Status.ContainerStatuses[i]
+		if sidecars.Has(status.Name) && status.State.Terminated == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func allMainContainersHaveBeenCompleted(pod *corev1.Pod) bool {
+	sidecars, _ := getSidecarNames(pod)
+	for i := range pod.Status.ContainerStatuses {
+		status := &pod.Status.ContainerStatuses[i]
+		if !sidecars.Has(status.Name) && status.State.Terminated == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func allMainContainersHaveBeenSucceeded(pod *corev1.Pod) bool {
+	sidecars, _ := getSidecarNames(pod)
+	for i := range pod.Status.ContainerStatuses {
+		status := &pod.Status.ContainerStatuses[i]
+		if !sidecars.Has(status.Name) &&
+			(status.State.Terminated == nil || status.State.Terminated.ExitCode != int32(0)) {
+			return false
+		}
+	}
+	return true
+}
+
+func updateImageForInPlaceUpdate(pod *corev1.Pod) *corev1.Pod {
+	sidecars, indexes := getSidecarNames(pod)
+	sidecars = filterOutUncompletedSidecars(pod, sidecars)
+	for _, i := range indexes {
+		container := &pod.Spec.Containers[i]
+		if sidecars.Has(container.Name) {
+			newImage := getImageFromEnv(container)
+			if len(newImage) == 0 {
```

Comment:
Is this too late? Should we have a pre-check in webhook or event handler or the beginning of reconcile that checks if sidecar containers with `INPLACE_UPDATE_ONLY` have `KRUISE_TERMINATE_SIDECAR_WHEN_JOB_EXIT_WITH_IMAGE` env or a global flag for default image?